### DISCUSSION TITLE
Running BepuPhysis in Unity With NativeAOT Interlop

### DISCUSSION THREAD

**Discussion by sayiliromer** [bepu/bepuphysics2#331] [Show and tell]
Along with Unity users, i beleive running BepuPhysics inside Unit was a dream.
NativeAOT allows you to expose functions to unmanaged callers. So those functions can be called from any platforms that supports C interlops. 
Here is a basic example by Mircosoft. https://github.com/dotnet/samples/tree/main/core/nativeaot/NativeLibrary

What i did is, i exposed some basic functions by wraping some BepuPhysics APIs.
I haven't done a performance comparison but i can tell it is reasonably good. Feels close to C# JIT version.
You can find a video of BepuPhysics runnind inside Unity at full speed. 
Also, i included an example code if you want to try something similar.

<div align="center">
      <a href="https://www.youtube.com/watch?v=ylW0vs2i9wU">
         <img src="https://img.youtube.com/vi/ylW0vs2i9wU/0.jpg" style="width:100%;">
      </a>
</div>

        [UnmanagedCallersOnly(EntryPoint = "CreateSimulationInstance")]
        public static int CreateSimulationInstance()
        {
            var sim = Simulation
                .Create(
                    _bufferPool,
                    new NarrowPhaseCallbacks(new SpringSettings(30, 1)),
                    new PoseIntegratorCallbacks(new Vector3(0, -10, 0)),
                    new SolveDescription(8, 1));
            var instance = new SimInstance(sim, _bufferPool);

            int index;
            if (_emptyIndexes.Count != 0)
            {
                index = _emptyIndexes.Dequeue();
                _instances[index] = instance;
            }
            else
            {
                index = _instances.Count;
                _instances.Add(instance);
            }

            return index;
        }
        [UnmanagedCallersOnly(EntryPoint = "DestroySimulation")]
        public static bool DestroySimulation(int simId)
        {
            if (simId >= _instances.Count) return false;
            var simInstance = _instances[simId];
            simInstance.Dispose();
            _emptyIndexes.Enqueue(simId);
            _instances[simId] = default;
            return true;
        }
        [UnmanagedCallersOnly(EntryPoint = "AddBody")]
        public static int AddBody(int simId, PhysicsTransform transform, Vector3 velocity, BodyInertiaData inertiaData, uint packedShape, float sleepThreshold)
        {
            var bd = new BodyDescription()
            {
                Pose = new RigidPose(transform.Position,transform.Rotation),
                Velocity = velocity,
                LocalInertia = inertiaData.ReinterpretCast<BodyInertiaData, BodyInertia>(),
                Collidable = new TypedIndex() { Packed = packedShape },
                Activity = new BodyActivityDescription(sleepThreshold)
            };
            return _instances[simId].Simulation.Bodies.Add(bd).Value;
        }

**Comment by RossNordby** [comment]
Should be full performance if nativeaot is allowed to use AVX2, yup! Notably, there is a still-incomplete C API of this sort over in the abomination interop project, plus a partial C++ wrapper:
https://github.com/RossNordby/scratchpad/tree/main/Abomination/AbominationInterop/AbominationInterop
